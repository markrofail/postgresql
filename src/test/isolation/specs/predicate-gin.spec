# Test for page level predicate locking in gin
#
# Test to verify serialization failures
#
# Queries are written in such a way that an index scan(from one transaction) and an index insert(from another transaction) will try to access the same part(sub-tree) of the index.


setup
{
 create table gin_tbl(id int4, p int4[]);
 create index ginidx on gin_tbl using gin(p) with 
 (fastupdate = off);
 insert into gin_tbl select g, array[1, 2, g*2] from generate_series(1, 200) g;
 insert into gin_tbl select g, array[3, 4, g*3] from generate_series(1, 200) g;
}

teardown
{
 DROP TABLE gin_tbl;
}

session "s1"
setup		{ 
		  BEGIN ISOLATION LEVEL SERIALIZABLE;
		  set enable_seqscan=off;
		}
step "rxy1"	{ select sum(p[2]) from gin_tbl where p @> array[1,2]; }
step "wx1"	{ insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;}
step "c1"	{ COMMIT; }

session "s2"
setup		{ 
		  BEGIN ISOLATION LEVEL SERIALIZABLE;
		  set enable_seqscan=off;
		}

step "rxy2"	{ select sum(p[2]) from gin_tbl where p @> array[3,4]; }
step "wy2"	{ insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g; }
step "c2"	{ COMMIT; }
