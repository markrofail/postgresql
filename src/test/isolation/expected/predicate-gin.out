Parsed test spec with 2 sessions

starting permutation: rxy1 wx1 c1 rxy2 wy2 c2
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step c1: COMMIT;
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

1004           
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step c2: COMMIT;

starting permutation: rxy1 wx1 rxy2 c1 wy2 c2
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step c1: COMMIT;
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
ERROR:  could not serialize access due to read/write dependencies among transactions
step c2: COMMIT;

starting permutation: rxy1 wx1 rxy2 wy2 c1 c2
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step c1: COMMIT;
step c2: COMMIT;
ERROR:  could not serialize access due to read/write dependencies among transactions

starting permutation: rxy1 wx1 rxy2 wy2 c2 c1
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step c2: COMMIT;
step c1: COMMIT;
ERROR:  could not serialize access due to read/write dependencies among transactions

starting permutation: rxy1 rxy2 wx1 c1 wy2 c2
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step c1: COMMIT;
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
ERROR:  could not serialize access due to read/write dependencies among transactions
step c2: COMMIT;

starting permutation: rxy1 rxy2 wx1 wy2 c1 c2
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step c1: COMMIT;
step c2: COMMIT;
ERROR:  could not serialize access due to read/write dependencies among transactions

starting permutation: rxy1 rxy2 wx1 wy2 c2 c1
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step c2: COMMIT;
step c1: COMMIT;
ERROR:  could not serialize access due to read/write dependencies among transactions

starting permutation: rxy1 rxy2 wy2 wx1 c1 c2
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step c1: COMMIT;
step c2: COMMIT;
ERROR:  could not serialize access due to read/write dependencies among transactions

starting permutation: rxy1 rxy2 wy2 wx1 c2 c1
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step c2: COMMIT;
step c1: COMMIT;
ERROR:  could not serialize access due to read/write dependencies among transactions

starting permutation: rxy1 rxy2 wy2 c2 wx1 c1
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step c2: COMMIT;
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
ERROR:  could not serialize access due to read/write dependencies among transactions
step c1: COMMIT;

starting permutation: rxy2 rxy1 wx1 c1 wy2 c2
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step c1: COMMIT;
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
ERROR:  could not serialize access due to read/write dependencies among transactions
step c2: COMMIT;

starting permutation: rxy2 rxy1 wx1 wy2 c1 c2
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step c1: COMMIT;
step c2: COMMIT;
ERROR:  could not serialize access due to read/write dependencies among transactions

starting permutation: rxy2 rxy1 wx1 wy2 c2 c1
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step c2: COMMIT;
step c1: COMMIT;
ERROR:  could not serialize access due to read/write dependencies among transactions

starting permutation: rxy2 rxy1 wy2 wx1 c1 c2
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step c1: COMMIT;
step c2: COMMIT;
ERROR:  could not serialize access due to read/write dependencies among transactions

starting permutation: rxy2 rxy1 wy2 wx1 c2 c1
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step c2: COMMIT;
step c1: COMMIT;
ERROR:  could not serialize access due to read/write dependencies among transactions

starting permutation: rxy2 rxy1 wy2 c2 wx1 c1
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step c2: COMMIT;
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
ERROR:  could not serialize access due to read/write dependencies among transactions
step c1: COMMIT;

starting permutation: rxy2 wy2 rxy1 wx1 c1 c2
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step c1: COMMIT;
step c2: COMMIT;
ERROR:  could not serialize access due to read/write dependencies among transactions

starting permutation: rxy2 wy2 rxy1 wx1 c2 c1
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step c2: COMMIT;
step c1: COMMIT;
ERROR:  could not serialize access due to read/write dependencies among transactions

starting permutation: rxy2 wy2 rxy1 c2 wx1 c1
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

400            
step c2: COMMIT;
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
ERROR:  could not serialize access due to read/write dependencies among transactions
step c1: COMMIT;

starting permutation: rxy2 wy2 c2 rxy1 wx1 c1
step rxy2: select sum(p[2]) from gin_tbl where p @> array[3,4];
sum            

800            
step wy2: insert into gin_tbl select g, array[1, 2, g*2] 
		  from generate_series(200, 250) g;
step c2: COMMIT;
step rxy1: select sum(p[2]) from gin_tbl where p @> array[1,2];
sum            

502            
step wx1: insert into gin_tbl select g, array[3, 4, g*3] 
		  from generate_series(200, 250) g;
step c1: COMMIT;
